### PR TITLE
Bug Fix + Hotkey Addition

### PR DESCRIPTION
**Pull request by ghost**
I'm new to GitHub and I was told by 05rHardy to make a pull request with my code. I'm not sure if this is the correct way to do things. I've added the ability to press z hotkey for reagent containers to change their transfer amounts and I've fixed issue #170, the poster transformation bug.

### CODE REVIEW

**Comment by Cruix** [comment]
Github is an enormous pain sometimes but it's much better than the alternative of manually merging everything.

Your pull request has made no changes, which you can see under the "files changed" tab. I suspect this is because you either didn't commit or didn't push. Commit your changes and push them to your repository, and this pull request should automatically update.

You are also using the master branch of your own fork, which isn't a problem for us, but could easily turn out to be a pain for you. I made the same mistake when I made my first pull request. Depending on what tool you are using to interact with Github, I may be able to help you fix it after we fix this pull request.

**Comment by ghost** [comment]
I tried to fix it by making a new branch and them committing/pushing to that one and now I don't have the right changes that I wanted to push.

**Comment by Cruix** [comment]
Are the changes still on your master branch?

**Comment by ghost** [comment]
I don't think so.

**Comment by ghost** [comment]
I had 8 uncommitted changes, now 6.

**Comment by Cruix** [comment]
Looks like the only commit pushed to your repository is a merge with the Yogstation dev branch. Make sure you are staging all of your changes before you commit.

**Comment by ghost** [comment]
I think I found the problem. I merged something or other with mine and I lost the changes I made. I'm fixing it now and I will commit it soon.
I'm using the GitHub desktop application thing.

**Comment by ghost** [comment]
I made a new commit with the two files I changed and nothing is happening.

**Comment by Cruix** [comment]
Did you push the commit?

If you did not, when you do, try to push it to a new be branch labeled "bugfix + hotkey addition" or something similar

**Comment by ghost** [comment]
Sorry, but, how do I push?

**Comment by ghost** [comment]
I've undone my last commit. You want me to make a new branch, commit then, and then what?

**Comment by Cruix** [comment]
Push should be a button near commit, it is what actually sends the commit to your repository. If you wait a few hours I can check out the Github desktop app and find it.

**Comment by ghost** [comment]
github

**Comment by Cruix** [comment]
Try the sync button in the upper right

**Comment by ghost** [comment]
It worked. Thank you so much.

**Comment by Cruix** [comment]
Nice job getting Git working. I'll have some feedback on your code in a few hours.

**Comment by ghost** [comment]
Thanks.

File: code/game/objects/effects/decals/contraband.dm [hunk 1]
```diff
@@ -369,8 +372,11 @@
 
 	if(istype(src,/turf/simulated/wall) && user && user.loc == temp_loc)	//Let's check if everything is still there
 		user << "<span class='notice'>You place the poster!</span>"
+	else if(istype(P, /obj/item/weapon/contraband/poster/legit))
+		D.roll_and_drop_legit(temp_loc)
 	else
 		D.roll_and_drop(temp_loc)
+	qdel(P)	//delete it now to cut down on sanity checks afterwards. Agouri's code supports rerolling it anyway //moved by Groxic to fix bug and check type
```
Comment (george99g) [inline code/game/objects/effects/decals/contraband.dm]: Don't leave your name in comments, please.
Git already manages name-tracking.

**Comment by Cruix** [comment]
Feedback:
First off, you should not make one pull request that adds a new feature and fixes an unrelated bug. You should make two separate branches and two separate pull requests. It may seem like a lot of trouble, but it helps if the repository owners want the bugfix but not the feature, or vice-versa.

I don't know why you changed line 82 in contraband.dm, but you should undo that. It won't change much, but there are some very subtle differences between the value 0 and the value null (which is what it is automatically assigned if you give it no default value).

The new roll_and_drop_legit() proc is good.

In the place_poster() proc, you did fix the bug. However, you moved the qdel from before the sleep() call to after it, so that you could check the type when you needed to. This leaves the poster in the hand of the person placing it, allowing them to place it multiple times before the first finishes unrolling. If the person does not move, only one of the posters is created on the wall, but if they do move, then two posters are dropped, which is a fairly serious duplication exploit. As in, if the clown gets one poster, soon the whole station will be covered in posters.

You can fix this by moving the qdel back to where it was, checking the type of P before the sleep() call, and storing the result in a variable that you look at when you need to.

In your reagent container code, you overrode the attack_self() proc twice, which you should almost never do. Furthermore, the only thing that the second attack_self() does is call the first one with the "..()" operator. You should remove the second attack_self() proc. (If you have experience in other object-oriented languages, you probably thought you were overriding two different overloaded procs, since they have different argument lists. However, DM does not care about the number or type of arguments when defining or calling a proc. Combined with DM's unorthadox mechanic of being able to override the same proc more than once, this can easily cause confusion.)

Because you used the attack_self() proc, people will be able to set the transfer amount not only with the hotkey, but also by selecting the reagent container in their active hand and clicking on it. This is fine, but you should make it more clear in your pull request - you emphasized the fact that it can be done with the hotkey, and did not mention that you could also do it by using the mouse.

In the first attack_self() proc, you clearly copy-pasted all the code from the set_APTFT() verb. While this worked, you should try to avoid copy-pasting code whenever possible, so that if the code is ever updated, you only have to update it in one place. Even in situations like this, where the copied and original code are right next to each other, you should still follow this convention.

You can fix this by just calling set_APTFT() in the attack_self() proc. Normally you would also have to add a range check, since the "set src in range(0)" line would not work when the verb is called by a proc like attack_self(), but because you can only call attack_self() when you are holding an item in your hand, there is no need for a range check. (With the exception of people using Telekenesis. If you care about people with TK being able to set the transfer amount of a beaker/bucket/etc. from long range, then add a range check or a contents check.)

Do not be discouraged. Though I described these issues in a very wordy way, these changes are very small, and for the most part your code does work as intended. By far the hardest part of coding is starting out, and the vast majority of people do not ever get to the point where they can even make a pull request.

**Comment by ghost** [comment]
Thank you. Do I make the changes and then commit again? I didn't realize verbs could be called within procs. I couldn't find the original code for using hotkeys and I only found attack_self().

**Comment by Cruix** [comment]
Yes, if you make changes, commit, and sync again this pull request should auto-update.

**Comment by ghost** [comment]
Done.

**Comment by ghost** [comment]
Also, I have another GitHub account, GRUBIK. Do you know why things, such as my merges/commits are going through that account?

**Comment by Cruix** [comment]
I have no idea, sorry.

**Comment by Cruix** [comment]
Code looks good, excellent work.

**Pull request merged**